### ISSUE TITLE
Problems at training OMR dataset

### ISSUE THREAD

**Issue by chengzhanzhan** [bgshih/crnn#9]
I tried to train crnn on dataset PitchRec_dataset\OMRB\TrainSet (2345 images), then got "train loss = inf" just like:

[05/23/16 11:18:15]  Loading datasets...
[05/23/16 11:18:15]  Start training...
[05/23/16 11:18:15]  Validating...
[05/23/16 11:18:17]  Test loss = 75.251866, accuracy = 0.000000
[05/23/16 11:18:17]  dddddddddddddddddddddddddd  =>  d                          (GT:aadee               )
[05/23/16 11:18:17]  dd1ddddddddddddddddddddddd  =>  d1d                        (GT:ghihhi              )
[05/23/16 11:18:17]  7ddddddddddddddddddddddddh  =>  7dh                        (GT:dcbdddfdfg          )
[05/23/16 11:18:17]  dd1dddd111111111111111111h  =>  d1d1h                      (GT:gfggfghhhj          )
[05/23/16 11:18:17]  dd11dddddddddddddddddddddd  =>  d1d                        (GT:ihfcba              )
[05/23/16 11:18:17]  711ddddddddddddddddddddddh  =>  71dh                       (GT:ihgh                )
[05/23/16 11:18:17]  d11111111111ddddd111111ddh  =>  d1d1dh                     (GT:fjjejjdjjc          )
[05/23/16 11:18:17]  7d1111dddddddddddd11dd111h  =>  7d1d1d1h                   (GT:gccbcchgg           )
[05/23/16 11:18:17]  dddddddddddddddddddddddddh  =>  dh                         (GT:adeef               )
[05/23/16 11:18:17]  dddddddddddddddddddddddddh  =>  dh                         (GT:fgfe                )
[05/23/16 11:18:17]  7d11ddddddddddddddddd1111h  =>  7d1d1h                     (GT:iffefhg             )
[05/23/16 11:18:17]  dddddddddddddddddddddddddd  =>  d                          (GT:edcb                )
[05/23/16 11:18:17]  dddddddddddddddddddddddd1h  =>  d1h                        (GT:fhhfhhcee           )
[05/23/16 11:18:17]  dddddddddddddddddddddddddh  =>  dh                         (GT:bccddded            )
[05/23/16 11:18:17]  dd1dddddddddddddd1111ddddh  =>  d1d1dh                     (GT:chi                 )
[05/23/16 11:18:47]  Iteration 100 - train loss = inf
[05/23/16 11:19:17]  Iteration 200 - train loss = inf
[05/23/16 11:19:47]  Iteration 300 - train loss = inf
[05/23/16 11:20:17]  Iteration 400 - train loss = inf
[05/23/16 11:20:47]  Iteration 500 - train loss = inf
[05/23/16 11:21:17]  Iteration 600 - train loss = inf
[05/23/16 11:21:45]  Iteration 700 - train loss = inf
[05/23/16 11:22:11]  Iteration 800 - train loss = inf
[05/23/16 11:22:37]  Iteration 900 - train loss = inf
[05/23/16 11:23:04]  Iteration 1000 - train loss = inf

My config.lua is as follows:
function getConfig()
    local config = {
        nClasses         = 36,
        maxT             = 26,
        displayInterval  = 100,
        testInterval     = 1000,
        nTestDisplay     = 15,
        trainBatchSize   = 64,
        valBatchSize     = 100,
        snapshotInterval = 1000,
        maxIterations    = 2000000,
        optimMethod      = optim.adadelta,
        optimConfig      = {},
        trainSetPath     = '../../PitchRec_dataset/LMDB/TrainSet/data.mdb',
        valSetPath       = '../../PitchRec_dataset/LMDB/Synthesized/data.mdb',
    }
    return config
end

function createModel(config)
    local nc = config.nClasses
    local nl = nc + 1
    local nt = config.maxT

    local ks = {3, 3, 3, 3, 3, 3, 2}
    local ps = {1, 1, 1, 1, 1, 1, 0}
    local ss = {1, 1, 1, 1, 1, 1, 1}
    local nm = {64, 128, 256, 256, 512, 512, 512}
    local nh = {256, 256}

    function convRelu(i, batchNormalization)
        batchNormalization = batchNormalization or false
        local nIn = nm[i-1] or 1
        local nOut = nm[i]
        local subModel = nn.Sequential()
        local conv = cudnn.SpatialConvolution(nIn, nOut, ks[i], ks[i], ss[i], ss[i], ps[i], ps[i])
        subModel:add(conv)
        if batchNormalization then
            subModel:add(nn.SpatialBatchNormalization(nOut))
        end
        subModel:add(cudnn.ReLU(true))
        return subModel
    end

    function bidirectionalLSTM(nIn, nHidden, nOut, maxT)
        local fwdLstm = nn.LstmLayer(nIn, nHidden, maxT, 0, false)
        local bwdLstm = nn.LstmLayer(nIn, nHidden, maxT, 0, true)
        local ct = nn.ConcatTable():add(fwdLstm):add(bwdLstm)
        local blstm = nn.Sequential():add(ct):add(nn.BiRnnJoin(nHidden, nOut, maxT))
        return blstm
    end

    -- model and criterion
    local model = nn.Sequential()
    model:add(nn.Copy('torch.ByteTensor', 'torch.CudaTensor', false, true))
    model:add(nn.AddConstant(-128.0))
    model:add(nn.MulConstant(1.0 / 128))
    model:add(convRelu(1))
    model:add(cudnn.SpatialMaxPooling(2, 2, 2, 2))       -- 64x16x50
    model:add(convRelu(2))
    model:add(cudnn.SpatialMaxPooling(2, 2, 2, 2))       -- 128x8x25
    model:add(convRelu(3, true))
    model:add(convRelu(4))
    model:add(cudnn.SpatialMaxPooling(2, 2, 1, 2, 1, 0)) -- 256x4x?
    model:add(convRelu(5, true))
    model:add(convRelu(6))
    model:add(cudnn.SpatialMaxPooling(2, 2, 1, 2, 1, 0)) -- 512x2x26
    model:add(convRelu(7, true))                         -- 512x1x26
    model:add(nn.View(512, -1):setNumInputDims(3))       -- 512x26
    model:add(nn.Transpose({2, 3}))                      -- 26x512
    model:add(nn.SplitTable(2, 3))
    model:add(bidirectionalLSTM(512, 256, 256, nt))
    model:add(bidirectionalLSTM(256, 256,  nl, nt))
    model:add(nn.SharedParallelTable(nn.LogSoftMax(), nt))
    model:add(nn.JoinTable(1, 1))
    model:add(nn.View(-1, nl):setNumInputDims(1))
    model:add(nn.Copy('torch.CudaTensor', 'torch.FloatTensor', false, true))
    model:cuda()
    local criterion = nn.CtcCriterion()

    return model, criterion
end

Then I tried to remove the 4th and 6th conv by referring to your paper (BiLSTMs are not replaced singleLSTM, because I'm new to torch. ), but it doesn't work.

Thanks.

**Comment by chengzhanzhan** [comment]
When I set "displayInterval  = 5", I found the overall process will converge; And some Iterations will show "train loss = inf", which may result from the optimMethod "adadelta".

**Comment by bgshih** [comment]
@chengzhanzhan Have you tried setting a larger `gConfig.maxT`? Besides, I have just fixed a bug in the code. You can check out the latest code and try again.

**Comment by chengzhanzhan** [comment]
@bgshih I have ever tried to set gConfig.maxT = 100, but a error occurs as follows:

[05/25/16 14:43:38]  Loading datasets...
[05/25/16 14:43:38]  Start training...
[05/25/16 14:43:41]  Validating...
/home/zzcheng/software/torch/install/bin/luajit: ./training.lua:48: inconsistent tensor size at /home/zzcheng/software/torch/pkg/torch/lib/TH/generic/THTensorCopy.c:7
stack traceback:
        [C]: in function 'copy'
        ./training.lua:48: in function 'validation'
        ./training.lua:85: in function 'trainModel'
        main_train.lua:51: in main chunk
        [C]: in function 'dofile'
        ...ware/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00406670

In your project, 26 is given to gConfig.maxT; I think that 26 should be the input length of LSTM. And when the network is defined, the gConfig.maxT is also decided.

Is it really like this ？

**Comment by bgshih** [comment]
@chengzhanzhan Sorry for the late response. Too busy recently. Have you already solved it? The code was designed to support variable length in training, if you look into the code (for example, https://github.com/bgshih/crnn/blob/master/src/LstmLayer.lua#L124), it actually supports any length <= maxT. You can check where the size inconsistency happens.

**Comment by chengzhanzhan** [comment]
@bgshih I have solved it. Thanks.

**Comment by oneOfThePeople** [comment]
hi @chengzhanzhan can you tell how you solved the problem?
Thanks

**Comment by Derongan** [comment]
@oneOfThePeople , you probably are way past this, but I ran into the same issue and fixed it by lowering the learning rate.

**Comment by oneOfThePeople** [comment]
thank a lot

**Comment by misssprite** [comment]
@oneOfThePeople 
My problem with similar symptom is caused by sequences with continuous characters, e.g., some id like '10000903'.
CTC inserts special 'blank' character between the chars in the sequence. To predict such a sequence, the model have to obtain a `path` like '1--0-0-0-0-90--3'.
This implies that the 'safe' length of the path should be at least `2n-1`, where n is the length of the sequence to predict.

So in my run, a sequence image with continuous chars will cause a inf loss, thus producing a inf batch loss. If you reduce the batch size, you'll luckily get some small batches without such sequences and obtain normal batch losses, and the training phase converges after all. So you'll see normal batches with inf loss in between.

The simplest hack to my problem is to enlarge the width of the input images to make the output path at least twice as the maximum length of sequences to predict.

Hope it help.

**Comment by srihari-humbarwadi** [comment]
@misssprite i was using CTC loss for text recognition, and your explanation makes so much sense! thanks sir!
